plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -61,227 +43,123 @@
-        this.metastore = requireNonNull(metastore, "metastore is null");
-        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
         this.typeManager = requireNonNull(typeManager, "typeManager is null");
         this.commitTaskCodec = requireNonNull(commitTaskCodec, "commitTaskCodec is null");
-        this.tableOperationsProvider = requireNonNull(tableOperationsProvider, "tableOperationsProvider is null");
-        this.trinoVersion = requireNonNull(trinoVersion, "trinoVersion is null");
+        this.catalog = requireNonNull(catalog, "catalog is null");
     }
 
     @Override
     public List<String> listSchemaNames(ConnectorSession session)
     {
-        return metastore.getAllDatabases().stream()
-                .filter(schemaName -> !HiveUtil.isHiveSystemSchema(schemaName))
-                .collect(toImmutableList());
-    }
-
-    private List<String> listSchemas(ConnectorSession session, Optional<String> schemaName)
-    {
-        if (schemaName.isPresent()) {
-            if (isHiveSystemSchema(schemaName.get())) {
-                return ImmutableList.of();
-            }
-            return ImmutableList.of(schemaName.get());
-        }
-        return listSchemaNames(session);
+        return catalog.listNamespaces(session);
     }
 
     @Override
     public Map<String, Object> getSchemaProperties(ConnectorSession session, CatalogSchemaName schemaName)
     {
-        Optional<Database> db = metastore.getDatabase(schemaName.getSchemaName());
-        if (db.isPresent()) {
-            return HiveSchemaProperties.fromDatabase(db.get());
-        }
-
-        throw new SchemaNotFoundException(schemaName.getSchemaName());
+        return catalog.loadNamespaceMetadata(session, schemaName.getSchemaName());
     }
 
     @Override
     public Optional<TrinoPrincipal> getSchemaOwner(ConnectorSession session, CatalogSchemaName schemaName)
     {
-        Optional<Database> database = metastore.getDatabase(schemaName.getSchemaName());
-        if (database.isPresent()) {
-            return database.flatMap(db -> Optional.of(new TrinoPrincipal(db.getOwnerType(), db.getOwnerName())));
-        }
-
-        throw new SchemaNotFoundException(schemaName.getSchemaName());
+        return catalog.getNamespacePrincipal(session, schemaName.getSchemaName());
     }
 
     @Override
     public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
     {
         IcebergTableName name = IcebergTableName.from(tableName.getTableName());
-        verify(name.getTableType() == DATA, "Wrong table type: " + name.getTableType());
+        verify(name.getTableType() == DATA, "Wrong table type: " + name.getTableNameWithType());
 
-        Optional<Table> hiveTable = metastore.getTable(new HiveIdentity(session), tableName.getSchemaName(), name.getTableName());
-        if (hiveTable.isEmpty()) {
+        if (!catalog.tableExists(session, tableName)) {
```

Comment:
We should handle this using `loadTable()` so that we don't call the catalog API twice

Reply:
This is to keep the current behavior of 4 metastore calls for select queries. I agree we can simplify this, but I feel it's too much change of behavior for a refactor PR and it's better to use another PR for it.

Reply:
I would normally agree that a performance improvement should be done separately, but since we're replacing all this code and defining a new API, it makes sense design the API to be efficient. 

It looks like `tableExists` is only used here, so if we remove it, then we don't need it in the API at all.

Reply:
Sure I have updated the code and all the metastore count in tests and removed `tableExists` API. please let me know if there is any other concern.